File: src/Library.java
import java.util.ArrayList;

public class Library {
    private ArrayList<User> users;
    private ArrayList<Book> books;

    public Library() {
        users = new ArrayList<>();
        books = new ArrayList<>();
    }

    public void registerUser(User user) {
        users.add(user);
    }

    public void registerBook(Book book) {
        books.add(book);
    }

    public void rentBook(String userName, String bookTitle) {
        User user = getUserByName(userName);
        Book book = getBookByTitle(bookTitle);
        if (user != null && book != null && book.getRenter() == null) {
            user.rentBook(book);
        } else {
            System.out.println("User or book not found, or book already rented.");
        }
    }

    public ArrayList<User> getAllUsers() {
        return users;
    }

    public ArrayList<Book> getAllBooks() {
        return books;
    }

    public ArrayList<User> getUsersWithRentedBooks() {
        ArrayList<User> usersWithRentedBooks = new ArrayList<>();
        for (User user : users) {
            if (!user.getRentedBooks().isEmpty()) {
                usersWithRentedBooks.add(user);
            }
        }
        return usersWithRentedBooks;
    }

    public ArrayList<Book> getUnrentedBooks() {
        ArrayList<Book> unrentedBooks = new ArrayList<>();
        for (Book book : books) {
            if (book.getRenter() == null) {
                unrentedBooks.add(book);
            }
        }
        return unrentedBooks;
    }

    public ArrayList<Book> getRentedBooks() {
        ArrayList<Book> rentedBooks = new ArrayList<>();
        for (Book book : books) {
            if (book.getRenter() != null) {
                rentedBooks.add(book);
            }
        }
        return rentedBooks;
    }

    private User getUserByName(String name) {
        for (User user : users) {
            if (user.getName().equals(name)) {
                return user;
            }
        }
        return null;
    }

    private Book getBookByTitle(String title) {
        for (Book book : books) {
            if (book.getTitle().equals(title)) {
                return book;
            }
        }
        return null;
    }
}

File: src/src/library.py
class Library:
    def __init__(self):
        self.users = []
        self.books = []

    def register_user(self, user):
        self.users.append(user)

    def register_book(self, book):
        self.books.append(book)

    def rent_book(self, user_name, book_title):
        user = self.get_user_by_name(user_name)
        book = self.get_book_by_title(book_title)
        if user and book and not book.get_renter():
            user.rent_book(book)
        else:
            print("User or book not found, or book already rented.")

    def get_all_users(self):
        return self.users

    def get_all_books(self):
        return self.books

    def get_users_with_rented_books(self):
        return [user for user in self.users if user.get_rented_books()]

    def get_unrented_books(self):
        return [book for book in self.books if not book.get_renter()]

    def get_rented_books(self):
        return [book for book in self.books if book.get_renter()]

    def get_user_by_name(self, name):
        for user in self.users:
            if user.get_name() == name:
                return user
        return None

    def get_book_by_title(self, title):
        for book in self.books:
            if book.get_title() == title:
                return book
        return None
